src/sol_binder/nonce/base.py

```diff
@@ -58,13 +62,34 @@ def sync_from_chain(self, new_accounts: List[HexAddress] = tuple()):
     def _sync_from_chain(self, account: HexAddress):
         return self._set(account, self.__w3.eth.get_transaction_count(account))
 
-    def get_and_increment(self, account: HexAddress) -> Nonce:
-        nonce = self._get_and_increment(account)
-        self._get_logger().debug(f"Nonce for {account}: {nonce}")
-        return nonce
+    @contextmanager
+    def _lock(self):
+        raise NotImplementedError
 
-    def _get_and_increment(self, account: HexAddress) -> Nonce:
-        raise NotImplementedError()
+    @contextmanager
+    def advance_nonce(self, account: HexAddress):
+        with self._lock():
+            if account in self._suspected_desync:
+                self._sync_from_chain(account)
+            current_nonce = self._get(account)
+            try:
+                yield current_nonce
+            except Exception as e:
+                exception_types = [type(exc) for exc in expand(lambda exc: exc.__context__, e)]
+                if ReadTimeout in exception_types:
+                    self._suspected_desync.add(account)
+                raise e
+            else:
+                self._set(account, current_nonce + 1)
+                self._suspected_desync.discard(account)
+
+    def _get(self, account: HexAddress):
+        """Get the next nonce to use"""
+        raise NotImplementedError
+
+    def _increment(self, account: HexAddress):
+        self._set(account, self._get(account) + 1)
 
     def _set(self, account: HexAddress, nonce: Nonce):
```

Comment:
If _set keeps being 'protected' and not 'private' it must use the lock in case someone else messes with it

The correct way to do this would be either
1. Removing _set altogether to a private method or losing it
2. Creating a private _set_unsafe (unlocked, unimplemented) and a __set (safe, locked) and then calling __set_unsafe from already-locked contexted